// @ghostery/ui/src/utils/iframe.js
/**
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2017-present Ghostery GmbH. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */


function closeIframe(reload = false, clear = false) {
  // In some cases await for store.set() is not enough to propagate changes
  // so we need to wait a bit before sending the message
  setTimeout(() => {
    window.parent.postMessage(
      { type: `ghostery-close-iframe`, reload, clear },
      '*',
    );
  }, 100);
}

function setupIframe(width) {
  const resizeObserver = new ResizeObserver(() => {
    window.parent.postMessage(
      {
        type: `ghostery-resize-iframe`,
        height: document.body.clientHeight,
        width,
      },
      '*',
    );
  });

  resizeObserver.observe(document.body, {
    box: 'border-box',
  });

  document.body.style.overflow = 'hidden';

  chrome.runtime.onMessage.addListener((msg) => {
    if (msg.action === 'clearIframe') {
      console.log('clearIframe', msg.url);
      window.parent.postMessage(
        {
          type: `ghostery-clear-iframe`,
          url: msg.url,
        },
        '*',
      );
    }
  });
}

export { closeIframe, setupIframe };
